frontend/src/layouts/Dapp.js

```diff
@@ -1,36 +1,79 @@
 import React, {useState, useEffect} from "react";
-import {authenticate} from "ceramic/index.js";
 import Sidebar from "components/Sidebar.js";
 import WalletInfo from "components/WalletInfo.js";
 
 import {useLocation} from "react-router-dom";
 
 import {DappContextProvider} from "context/dappContext";
 
+import {getServerDID} from "api/index";
+
+import {readOnlyClient, authenticatedClient} from "identity";
+import {connect} from "ethereum";
+
 const Dapp = ({children}) => {
-  /* Ceramic Code */
+  /* --- Ceramic Code --- */
+
+  // Read Only Ceramic + IDX
+  const [readOnlyClients, setReadOnlyClients] = useState({});
+  async function getReadOnlyClients() {
+    const {ceramic, idx} = await readOnlyClient();
+    setReadOnlyClients((prevState) => {
+      return {...prevState, ...{ceramic, idx}};
+    });
+  }
+
+  // getReadOnly instance on mounted
+  useEffect(() => {
+    getReadOnlyClients();
+  }, []);
+
+  // Authenticated Ceramic + IDX + DID
+  const [authenticatedClients, setAuthenticatedClients] = useState({});
+  async function getAuthenticatedClients() {
+    const {idx, ceramic, did} = await authenticatedClient();
+    setAuthenticatedClients((prevState) => {
+      return {...prevState, ...{idx, ceramic, did}};
```

Comment:
Why do we need the prevState ? Is it only to get the `error` field ?

Reply:
From what I understand, and I might be wrong, this is made so that we can use objects in `useState`, because if we do a regular assignment `=` React has no way of knowing if there were changes because the object is passed by reference, thus we need to do that, or `Object.assign({}, prevState, {idx, ceramic, did})`
But in this case it might not be necessary because we're only running it once! Not sure to be honest.